### PR TITLE
Cli

### PR DESCRIPTION
**Pull request by Arnatious**
add cli with two verbs
- `show`: display either raw or parsed dict of nodl data
- `validate`: attempt to validate a file against the schema

### CODE REVIEW

File: ros2nodl/ros2nodl/verb/show.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    def add_arguments(self, parser: 'argparse.ArgumentParser', *_):
+        parser.add_argument(  # type: ignore
+            'package_name', help='Name of the package to show.'
+        ).completer = package_name_completer
+
+        command_group = parser.add_mutually_exclusive_group(required=True)
+        command_group.add_argument(
+            '-a',
+            '--all',
+            '--all-files',
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/show.py:42]: I don't know that `--all-files` adds any meaning beyond `--all`. I suggest eliminating it.

File: ros2nodl/ros2nodl/verb/show.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            default=False,
+            action='store_true',
+            help='Combine all .nodl.xml files and display the combined output.',
+        )
+        command_group.add_argument(  # type: ignore
+            '-f', '--file', nargs='+', help='Specific .nodl.xml file(s) to display.'
+        ).completer = NoDLFileNameCompleter()
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/show.py:49]: I would have expected this to be executable name rather than a file path. This route has some assumed knowledge (that nodl files exist, where they're installed, etc.). What is the reason we're doing it this way?

File: ros2nodl/ros2nodl/verb/validate.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from pathlib import Path
+import sys
+from typing import TYPE_CHECKING
+
+from argcomplete.completers import FilesCompleter
+import nodl
+from ros2nodl.verb import VerbExtension
+
+if TYPE_CHECKING:  # pragma: no cover
+    import argparse
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/validate.py:22]: Ditto previous comments.

File: ros2nodl/ros2nodl/verb/validate.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class ValidateVerb(VerbExtension):
+    """Validate NoDL XML documents."""
+
+    def add_arguments(self, parser: 'argparse.ArgumentParser', *_):  # pragma: no cover
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/validate.py:28]: Ditto previous comment about `*_`. Also, why the pragma here?

File: ros2nodl/ros2nodl/verb/validate.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        command_group.add_argument(
+            '-a',
+            '--all',
+            '--all-files',
+            default=False,
+            action='store_true',
+            help='Validate all .nodl.xml files in the current directory.',
+        )
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        # TODO: remove type ignore after Focal/Foxy release
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/validate.py:39]: Ah, can you please add this to the [other verb](https://github.com/ubuntu-robotics/nodl/pull/9/files#diff-d26363672dfbffa78f0230740f5d3a87R34)? I was wondering about that.

File: ros2nodl/ros2nodl/verb/validate.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        command_group.add_argument(  # type: ignore
+            'file', nargs='*', default=[], help='Specific .nodl.xml file(s) to validate.'
+        ).completer = FilesCompleter(allowednames=['nodl.xml'], directories=False)
+
+    def main(self, *, args: 'argparse.Namespace') -> int:
+        paths = (
+            list(Path.cwd().glob('*.nodl.xml'))
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/validate.py:46]: Now I'm doubly questioning the utility of having [this function](https://github.com/ubuntu-robotics/nodl/blob/master/nodl_python/nodl/_index.py#L25-L27). The reason given for having it was so that we wouldn't need to duplicate this glob everywhere, but we're doing it anyway. Can we either make sure we use it or get rid of it, please?

File: ros2nodl/setup.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        ('share/ament_index/resource_index/packages',
+            ['resource/' + package_name]),
+        ('share/' + package_name, ['package.xml']),
+    ],
+    install_requires=['setuptools'],
+    zip_safe=True,
+    author='Ted Kern',
+    author_email='ted.kern@canonical.com',
+    maintainer='Ubuntu Robotics',
+    maintainer_email='ubuntu-robotics@lists.launchpad.net',
```
Comment (kyrofa) [inline ros2nodl/setup.py:19]: Let's just have the author instead of both author and maintainer, and leave it as Ubuntu Robotics, please. I think nodl is that way right now as well-- can you update it too, please?

File: ros2nodl/setup.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    description='TODO: Package description',
+    license='TODO: License declaration',
```
Comment (kyrofa) [inline ros2nodl/setup.py:21]: Missing some pieces here.

File: ros2nodl/ros2nodl/api/api.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    nodes = nodl._parsing._parsing._parse_multiple(paths=paths)
+    for node in nodes:
+        pprint.pprint(node._as_dict)
+
+
+class NoDLFileNameCompleter:
+    """Callable returning a list of nodl file names within a package's share directory."""
+
+    def __init__(self, *, package_name_key: Optional[str] = None) -> None:
+        self.package_name_key = 'package_name' if package_name_key is None else package_name_key
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py:54]: Was copied from https://github.com/ros2/launch_ros/blob/2e64ea5c0d6472543417444513827f696210d877/ros2launch/ros2launch/api/api.py#L167,
 will fix

File: ros2nodl/ros2nodl/verb/__init__.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    The following methods must be defined:
+    * `main`
+
+    The following methods can be defined:
+    * `add_arguments`
+    """
+
+    NAME = None
+    EXTENSION_POINT_VERSION = '0.1'
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py:30]: Borrowed logic from https://github.com/ros2/sros2/blob/master/sros2/sros2/verb/__init__.py

File: ros2nodl/ros2nodl/verb/__init__.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    def __init__(self):
+        super().__init__()
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py:33]: Again, borrowed logic from https://github.com/ros2/sros2/blob/master/sros2/sros2/verb/__init__.py
Will fix

File: ros2nodl/ros2nodl/verb/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py:30]: Preesumably replicated from https://github.com/ros2/ros2cli/blob/master/ros2cli/ros2cli/verb/__init__.py

File: ros2nodl/setup.py (hunk 8 above)
Comment (Arnatious) [reply ros2nodl/setup.py]: Fixed in b8c715a

File: ros2nodl/setup.py (hunk 7 above)
Comment (Arnatious) [reply ros2nodl/setup.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/validate.py (hunk 6 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/validate.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/validate.py (hunk 5 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/validate.py]: Fixed in b8c715a 
Had to also remove the "ignore after Focal/Foxy release" since Action is missing a completer attribute

File: ros2nodl/ros2nodl/api/api.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import pprint
+from typing import List, Optional, Sequence, TYPE_CHECKING
+
+import ament_index_python
+import nodl
+import nodl.errors
+
+if TYPE_CHECKING:  # pragma: no cover
+    import nodl.types
+    import argparse
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/api/api.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
+# Copyright 2020 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it under the terms of the
+# GNU Limited General Public License version 3, as published by the Free Software Foundation.
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/api/api.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+import ament_index_python
+import nodl
+import nodl.errors
+
+if TYPE_CHECKING:  # pragma: no cover
+    import nodl.types
+    import argparse
+
+
+def show_nodl_raw(*, paths: Sequence[Path]):
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/command/nodl.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+#
+# You should have received a copy of the GNU Limited General Public License along with
+# this program. If not, see <http://www.gnu.org/licenses/>.
+
+from typing import List, Optional, TYPE_CHECKING
+
+from ros2cli.command import add_subparsers_on_demand, CommandExtension
+
+if TYPE_CHECKING:
+    import argparse
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/command/nodl.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/api/api.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    """Print filename and contents to console.
+
+    :param paths: Names of nodl files to show
+    :type paths: Sequence[Path]
+    """
+    for path in paths:
+        print(f'{path.name}:\n' + ('-' * 80) + '\n')
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a, used get_terminal_size for pprint width and changed to whitespace

File: ros2nodl/ros2nodl/api/api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/api/api.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    def __init__(self, *, package_name_key: Optional[str] = None) -> None:
+        self.package_name_key = 'package_name' if package_name_key is None else package_name_key
+
+    def __call__(self, *, parsed_args: 'argparse.Namespace', **kwargs) -> List[str]:
+        """Return a list of file names for nodl files found within the package.
+
+        :param prefix: [description]
+        :type prefix: str
+        :param parsed_args: [description]
+        :type parsed_args: [type]
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/__init__.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Limited General Public License for more details.
+#
+# You should have received a copy of the GNU Limited General Public License along with
+# this program. If not, see <http://www.gnu.org/licenses/>.
+
+from ros2cli.plugin_system import PLUGIN_SYSTEM_VERSION, satisfies_version
+
+
+class VerbExtension:  # pragma: no cover
+    """The extension point for 'security' verb extensions.
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py]: Fixed in b8c715a, now inherits from ros2cli.VerbExtension

File: ros2nodl/ros2nodl/verb/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/show.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+from ros2nodl.api import (
+    NoDLFileNameCompleter,
+    show_nodl_parsed,
+    show_nodl_raw,
+)
+from ros2nodl.verb import VerbExtension
+from ros2pkg.api import package_name_completer, PackageNotFoundError
+
+if TYPE_CHECKING:  # pragma: no cover
+    import argparse
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/show.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/show.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class ShowVerb(VerbExtension):
+    """Show NoDL data."""
+
+    def add_arguments(self, parser: 'argparse.ArgumentParser', *_):
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/show.py]: Fixed in b8c715a

[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/show.py]: cli_name isn't used, by the way

File: ros2nodl/ros2nodl/verb/__init__.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    """
+
+    NAME = None
+    EXTENSION_POINT_VERSION = '0.1'
+
+    def __init__(self):
+        super().__init__()
+        satisfies_version(PLUGIN_SYSTEM_VERSION, '^0.1')
+
+    def add_arguments(self, parser, cli_name):
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/__init__.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/validate.py (hunk 3 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/validate.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/validate.py (hunk 4 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/validate.py]: Fixed in b8c715a

File: ros2nodl/ros2nodl/api/api.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
+# Copyright 2020 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/ros2nodl/api/api.py]: Fixed in b8c715a

File: ros2nodl/package.xml [hunk 23]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>ros2nodl</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="ted.kern@canonical.com">ubuntu</maintainer>
+  <license>TODO: License declaration</license>
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/package.xml]: Fixed in b8c715a

File: ros2nodl/package.xml [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+  <depend>argcomplete</depend>
+  <depend>launch</depend>
+  <depend>launch_ros</depend>
+  <depend>nodl_python</depend>
+  <depend>ros2cli</depend>
+  <depend>ros2pkg</depend>
+
+  <test_depend>ament_copyright</test_depend>
+  <test_depend>ament_flake8</test_depend>
+  <test_depend>ament_pep257</test_depend>
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply ros2nodl/package.xml]: Fixed in b8c715a

File: ros2nodl/ros2nodl/verb/show.py (hunk 2 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/show.py]: Fixed in 62602dd

File: ros2nodl/ros2nodl/verb/show.py (hunk 1 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/show.py]: Fixed in 62602dd

File: nodl_python/nodl/errors.py [hunk 25]
```diff
@@ -47,10 +47,19 @@ class InvalidNoDLError(NoDLError):
     """Exception class representing most errors in parsing the NoDL tree."""
 
 
+class InvalidXMLError(InvalidNoDLError):
+    """Error raised when unable to parse XML."""
+
+    def __init__(self, err: etree.XMLSyntaxError):
+        super().__init__(
+            f'XML Syntax Error: {err.filename}: {err.msg}'
```
Comment (kyrofa) [inline nodl_python/nodl/errors.py:55]:
```suggestion
            f'XML syntax error: {err.filename}: {err.msg}'
```

File: ros2nodl/package.xml [hunk 26]
```diff
@@ -2,10 +2,10 @@
 <?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
 <package format="3">
   <name>ros2nodl</name>
-  <version>0.0.0</version>
-  <description>TODO: Package description</description>
-  <maintainer email="ted.kern@canonical.com">ubuntu</maintainer>
-  <license>TODO: License declaration</license>
+  <version>0.0.1</version>
+  <description>CLI tools for NoDL files.</description>
+  <maintainer email="ubuntu-robotics@lists.launchpad.net">ubuntu</maintainer>
```
Comment (kyrofa) [inline ros2nodl/package.xml:7]:
```suggestion
  <maintainer email="ubuntu-robotics@lists.launchpad.net">Ubuntu Robotics</maintainer>
```

File: ros2nodl/test/ros2nodl/verbs/test_show.py [hunk 27]
```diff
@@ -1,40 +1,47 @@
...
+from typing import List
 
+from ament_index_python import PackageNotFoundError
 import nodl
 import pytest
+from ros2nodl.verb import show
+
 
-from ros2nodl.verb.show import PackageNotFoundError, ShowVerb
+sv = show.ShowVerb()
```
Comment (kyrofa) [inline ros2nodl/test/ros2nodl/verbs/test_show.py:25]: Maintaining state outside of individual tests is asking for trouble. Is there a reason this is required? pytest supports module-level fixtures if necessary, but this feels like something that could just be created within each test, too.

File: ros2nodl/ros2nodl/verb/validate.py [hunk 28]
```diff
@@ -1,49 +1,49 @@
-# Copyright 2020 Canonical Ltd.
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/validate.py]: This is still public API. Does it need to be?

File: ros2nodl/ros2nodl/verb/show.py [hunk 29]
```diff
@@ -1,82 +1,72 @@
-# Copyright 2020 Canonical Ltd.
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/show.py]: This is still public API. Does it need to be?

File: ros2nodl/ros2nodl/command/nodl.py [hunk 30]
```diff
@@ -1,21 +1,21 @@
-# Copyright 2020 Canonical Ltd.
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/command/nodl.py]: This is still public API. Does it need to be?

File: ros2nodl/ros2nodl/verb/validate.py [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+        )
+
+        parser.add_argument('-p', '--print', action='store_true', help='Print parsed output.')
+
+    def main(self, args: argparse.Namespace) -> int:
+        paths = (
+            list(nodl._index._get_nodl_xml_files_in_path(path=Path.cwd()))
+            if args.all
+            else [Path(filename) for filename in args.file]
+        )
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/verb/validate.py:49]: This logic is hard to follow across multiple lines. Let's use a normal if/else here, please, which uses the same number of lines but is way more clear:

```suggestion
        if args.all:
            paths = nodl._index._get_nodl_xml_files_in_path(path=Path.cwd())
        else:
            paths = [Path(filename) for filename in args.file]
```

Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/validate.py]: fixed in 48641b3

File: ros2nodl/package.xml (hunk 26 above)
Comment (Arnatious) [reply ros2nodl/package.xml]: fixed in 48641b3

File: nodl_python/nodl/errors.py (hunk 25 above)
Comment (Arnatious) [reply nodl_python/nodl/errors.py]: fixed in 48641b3

File: ros2nodl/ros2nodl/command/nodl.py (hunk 30 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/command/nodl.py]: fixed in 70d6ba5

File: ros2nodl/ros2nodl/verb/show.py (hunk 29 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/show.py]: fixed in 70d6ba5

File: ros2nodl/ros2nodl/verb/validate.py (hunk 28 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/verb/validate.py]: fixed in 70d6ba5

File: ros2nodl/test/ros2nodl/verbs/test_show.py (hunk 27 above)
Comment (Arnatious) [reply ros2nodl/test/ros2nodl/verbs/test_show.py]: fixed in 70d6ba5

File: nodl_python/nodl/errors.py [hunk 32]
```diff
@@ -47,10 +47,19 @@ class InvalidNoDLError(NoDLError):
...
+        super().__init__(
+            f'XML syntax error: {err.filename}: {err.msg}'
+        )
+
+
 class InvalidNoDLDocumentError(InvalidNoDLError):
     """Error raised when schema validation fails."""
 
-    def __init__(self, invalid: etree.DocumentInvalid) -> None:
+    def __init__(self, invalid: etree.DocumentInvalid):
```
Comment (kyrofa) [inline nodl_python/nodl/errors.py:62]: Why remove the `-> None`?

File: ros2nodl/ros2nodl/_verb/_show.py [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+        )
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        command_group.add_argument(  # type: ignore
+            'file', nargs='*', default=[], help='Specific Executable to display.'
+        ).completer = ExecutableNameCompleter(package_name_key='package_name')
+
+    def main(self, args: argparse.Namespace) -> int:
+        package = args.package_name
+        try:
+            nodes = nodl._index._get_nodes_from_package(package_name=package)
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_show.py:47]: It seems inefficient to do this before we know we need this below, no?

File: ros2nodl/ros2nodl/_verb/_show.py [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+    def add_arguments(self, parser: argparse.ArgumentParser, cli_name: None = None):
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        parser.add_argument(  # type: ignore
+            'package_name', help='Name of the package to show.'
+        ).completer = package_name_completer
+
+        command_group = parser.add_mutually_exclusive_group(required=True)
+        command_group.add_argument(
+            '-a', '--all', action='store_true', help='Show all executables in package.',
+        )
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_show.py:38]: I'm second-guessing this arg. What do we expect to happen if we say `ros2 nodl show <package>`? Right now that's an error, because either `-a` or `<executable>` is required. If we get rid of `-a` entirely and make the executable optional, we can default to the `-a` behavior if an executable wasn't specified.

File: ros2nodl/ros2nodl/_verb/_show.py [hunk 35]
```diff
@@ -0,0 +1,72 @@
...
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        command_group.add_argument(  # type: ignore
+            'file', nargs='*', default=[], help='Specific Executable to display.'
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_show.py:41]:
```suggestion
            'executable', nargs='*', default=[], help='Specific Executable to display.'
```

File: ros2nodl/ros2nodl/_verb/_show.py [hunk 36]
```diff
@@ -0,0 +1,72 @@
...
+        ).completer = ExecutableNameCompleter(package_name_key='package_name')
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_show.py:42]: Good call on this.

File: ros2nodl/ros2nodl/_verb/_validate.py [hunk 37]
```diff
@@ -0,0 +1,71 @@
...
+        command_group = parser.add_mutually_exclusive_group(required=True)
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        command_group.add_argument(  # type: ignore
+            'file', nargs='*', default=[], help='Specific .nodl.xml file(s) to validate.'
+        ).completer = FilesCompleter(allowednames=['nodl.xml'], directories=False)
+        command_group.add_argument(
+            '-a',
+            '--all',
+            action='store_true',
+            help='Validate all .nodl.xml files in the current directory.',
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_validate.py:39]: Similar to above, I'm not sure this arg is necessary. Thoughts?

File: ros2nodl/setup.py [hunk 38]
```diff
@@ -0,0 +1,41 @@
...
+    tests_require=['pytest'],
+    entry_points={
+        'console_scripts': [
+        ],
+        'ros2cli.command': [
+            'nodl = ros2nodl.command.nodl:NoDLCommand',
+        ],
+        'ros2nodl.verb': [
+            'show = ros2nodl.verb.show:ShowVerb',
+            'validate = ros2nodl.verb.validate:ValidateVerb'
```
Comment (kyrofa) [inline ros2nodl/setup.py:38]: You have moved these to non-public, this probably won't work anymore. Make sure you test this.

File: ros2nodl/ros2nodl/_verb/_show.py (hunk 34 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_show.py:38]: I don't like to assume, when run, that it should grab all nodl files, especially given it searches recursively. That being said, that's fairly standard for CLI tools. I'll change it to -a by default

File: ros2nodl/setup.py (hunk 38 above)
Comment (Arnatious) [reply ros2nodl/setup.py]: Fixed in 1b16352 , not sure if it was even causing any behavior though...

File: ros2nodl/ros2nodl/_verb/_validate.py (hunk 37 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_validate.py]: Fixed in 1b16352

File: ros2nodl/ros2nodl/_verb/_show.py (hunk 35 above)
Comment (kyrofa) [reply ros2nodl/ros2nodl/_verb/_show.py:37]: This is still relevant, no?

File: ros2nodl/ros2nodl/_command/_nodl.py [hunk 39]
```diff
@@ -0,0 +1,43 @@
...
+from typing import List, Optional
+
+from ros2cli.command import add_subparsers_on_demand, CommandExtension
+
+
+class _NoDLCommand(CommandExtension):
+    """Access node interface descriptions exported from packages."""
+
+    def add_arguments(
+        self, parser: 'argparse.ArgumentParser', cli_name: str, *, argv: Optional[List] = None
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_command/_nodl.py:25]: No need to quote `'argparse.ArgumentParser'`, now.

File: nodl_python/nodl/_index.py [hunk 40]
```diff
@@ -65,7 +65,26 @@ def get_node_by_executable(*, package_name: str, executable_name: str) -> Node:
...
     nodes = _get_nodes_from_package(package_name=package_name)
     try:
-        result = next(node for node in nodes if node.executable == executable_name)
+        return next(node for node in nodes if node.executable == executable_name)
     except StopIteration:
         raise ExecutableNotFoundError(package_name=package_name, executable_name=executable_name)
-    return result
+
+
+def get_nodes_by_executables(
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply nodl_python/nodl/_index.py:73]: If you're requesting multiple nodes, you'll want to use this rather than re-parse every time in the singular method. It could go in ros2nodl as a helper function but I can see others wanting to use it. I can remove it if need be.

[the remark this answers is not in the document]
Comment (kyrofa) [reply nodl_python/nodl/_index.py:73]: Let's wait and see if it's needed beyond the CLI. It's easy to make something public later, not so easy to make something non-public later.

File: ros2nodl/ros2nodl/_command/_nodl.py (hunk 39 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_command/_nodl.py]: fixed in 1d45ece

File: nodl_python/test/nodl/_parsing/test_nodl_parse.py [hunk 41]
```diff
@@ -68,8 +75,8 @@ def test_parse_multiple_accepts_disjoint_nodes(mocker):
     parse_mock = mocker.patch('nodl._parsing._parsing.parse')
 
     parse_mock.side_effect = [
-        [nodl.types.Node(name='foo', executable='foo')],
-        [nodl.types.Node(name='bar', executable='bar')],
+        {'foo': nodl.types.Node(name='foo', executable='foo')},
+        {'bar': nodl.types.Node(name='bar', executable='bar')},
```
[the remark this answers is not in the document]
Comment (Arnatious) [reply nodl_python/test/nodl/_parsing/test_nodl_parse.py]: 1d45ece

File: nodl_python/nodl/_index.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (Arnatious) [reply nodl_python/nodl/_index.py]: fixed in 1d45ece

File: ros2nodl/ros2nodl/_verb/_show.py (hunk 35 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_show.py]: fixed in 1d45ece

File: ros2nodl/ros2nodl/_verb/_show.py (hunk 33 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_show.py]: Fixed in 1d45ece

File: ros2nodl/ros2nodl/_verb/_validate.py [hunk 42]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class _ValidateVerb(VerbExtension):
+    """Validate NoDL XML documents."""
+
+    def add_arguments(self, parser: argparse.ArgumentParser, cli_name: None = None):
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        parser.add_argument(  # type: ignore
+            'executables', nargs='*', default=[], help='Specific .nodl.xml file(s) to validate.'
+        ).completer = FilesCompleter(allowednames=['nodl.xml'], directories=False)
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_validate.py:33]: This is duplicating the pattern maintained [here](https://github.com/ubuntu-robotics/nodl/blob/master/nodl_python/nodl/_index.py#L25). I don't know how else you'd do this, but it undermines the existence of `_get_nodl_xml_files_in_path`. Take stock of where you're using each, maybe it's time to get rid of it.

File: ros2nodl/ros2nodl/_verb/_show.py [hunk 43]
```diff
@@ -0,0 +1,65 @@
...
+
+    def add_arguments(self, parser: argparse.ArgumentParser, cli_name: None = None):
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        parser.add_argument(  # type: ignore
+            'package_name', help='Name of the package to show.'
+        ).completer = package_name_completer
+
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        parser.add_argument(  # type: ignore
+            'executables', nargs='*', default=[], help='Specific Executable to display.'
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_show.py:37]: Check your --help output, I think this will show up as `[executables [executables...]]` or similar. If so you might consider specifying the `metavar` arg here.

File: nodl_python/nodl/_index.py [hunk 44]
```diff
@@ -34,7 +32,7 @@ def _get_nodl_files_from_package_share(*, package_name: str) -> List[Path]:
     :raises NoNoDLFilesError: if no .nodl.xml files are in package share directory
     """
     package_share_directory = Path(get_package_share_directory(package_name))
-    nodl_paths = _get_nodl_xml_files_in_path(path=package_share_directory)
+    nodl_paths = [path for path in package_share_directory.glob('*.nodl.xml') if path.is_file()]
```
Comment (kyrofa) [inline nodl_python/nodl/_index.py:35]: Should this be using `_FILE_EXTENSION`?

File: ros2nodl/ros2nodl/_verb/_validate.py [hunk 45]
```diff
@@ -0,0 +1,69 @@
...
+    """Validate NoDL XML documents."""
+
+    def add_arguments(self, parser: argparse.ArgumentParser, cli_name: None = None):
+        # Ignoring type because of https://github.com/python/typeshed/issues/1878
+        parser.add_argument(  # type: ignore
+            'executables',
+            nargs='*',
+            default=[],
+            metavar='executable',
+            help='Specific .nodl.xml file(s) to validate.',
```
Comment (kyrofa) [inline ros2nodl/ros2nodl/_verb/_validate.py:37]: Would it make sense to use `_FILE_EXTENSION` here as well?

File: ros2nodl/ros2nodl/_verb/_validate.py (hunk 42 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_validate.py]: fixed in 0c6cd25

File: ros2nodl/ros2nodl/_verb/_show.py (hunk 43 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_show.py:37]: fixed in 0c6cd25

File: ros2nodl/ros2nodl/_verb/_validate.py (hunk 45 above)
Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_validate.py]: fixed in 3955cc9

File: nodl_python/nodl/_index.py (hunk 44 above)
Comment (Arnatious) [reply nodl_python/nodl/_index.py]: fixed in 3955cc9

File: ros2nodl/ros2nodl/_verb/_show.py (hunk 43 above)
Comment (kyrofa) [reply ros2nodl/ros2nodl/_verb/_show.py:37]: Hmm, was that commit lost?

Comment (Arnatious) [reply ros2nodl/ros2nodl/_verb/_show.py:37]: Ah  I subbed it in validate not here
